### PR TITLE
Fix ControlMaster behaviour

### PR DESCRIPTION
**Pull request by balsoft**
# Motivation

https://github.com/NixOS/nix/pull/8018 broke `ssh://` and `ssh-ng://` when `ControlMaster` is enabled. Fix it back.

# Context

Closes https://github.com/NixOS/nix/issues/8329#issuecomment-1549559917

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [x] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

File: src/libstore/ssh.cc [hunk 1]
```diff
@@ -152,13 +159,15 @@ Path SSHMaster::startMaster()
 
     out.writeSide = -1;
 
-    std::string reply;
-    try {
-        reply = readLine(out.readSide.get());
-    } catch (EndOfFile & e) { }
+    if (wasMasterRunning) {
```
Comment (thufschmitt) [inline src/libstore/ssh.cc:162]: Shouldn't that be the other way around?

```suggestion
    if (!wasMasterRunning) {
```

(maybe we could even make this whole function bail out early if `isMasterRunning() == true`, but let's save that for another time)

**Comment by Ericson2314** [comment]
Is it possible to test this? Maybe with a NixOS install test if it's hard with the functional tests?

**Review by max-privatevoid — APPROVED** [review]
Works with my manual test case

**Comment by Ericson2314** [comment]
(to be clear, I meant automatic tests to avoid regressions. I am perfectly willing to believe it works *today* :).)

**Comment by balsoft** [comment]
@Ericson2314 added a tests, sadly CI is flaky, waiting for a restart

**Review by Ericson2314 — APPROVED** [review]

**Pull request merged**